pytket/phir/phirgen.py

```diff
@@ -218,13 +218,110 @@ def convert_gate(op: tk.Op, cmd: tk.Command) -> JsonDict | None:
     return qop
 
 
+def cop_from_op_name(op_name: str) -> str:
+    """Get PHIR classical op name from pytket op name."""
+    match op_name:
+        case "AND":
+            cop = "&"
+        case "OR":
+            cop = "|"
+        case "XOR":
+            cop = "^"
+        case "NOT":
+            cop = "~"
+        case name:
+            raise NotImplementedError(name)
+    return cop
+
+
+def convert_classicalevalop(op: tk.ClassicalEvalOp, cmd: tk.Command) -> JsonDict | None:
+    """Return PHIR dict for a pytket ClassicalEvalOp."""
+    # Exclude conditional bit from args
+    args = cmd.args[1:] if isinstance(cmd.op, tk.Conditional) else cmd.args
+    out: JsonDict | None = None
+    match op:
+        case tk.CopyBitsOp():
+            if len(cmd.bits) != len(cmd.args) // 2:
+                logger.warning("LHS and RHS lengths mismatch for CopyBits")
+            out = assign_cop(
+                [arg_to_bit(bit) for bit in cmd.bits],
+                [arg_to_bit(args[i]) for i in range(len(cmd.args) // 2)],
+            )
+        case tk.SetBitsOp():
+            if len(cmd.bits) != len(op.values):
+                logger.error("LHS and RHS lengths mismatch for classical assignment")
+                raise ValueError
+            out = assign_cop(
+                [arg_to_bit(bit) for bit in cmd.bits], list(map(int, op.values))
+            )
+        case tk.RangePredicateOp():  # where the condition is a range
+            cond: JsonDict
+            match op.lower, op.upper:
+                case l, u if l == u:
+                    cond = {
+                        "cop": "==",
+                        "args": [cmd.args[0].reg_name, u],
+                    }
+                case l, u if u == UINTMAX:
+                    cond = {
+                        "cop": ">=",
+                        "args": [cmd.args[0].reg_name, l],
+                    }
+                case 0, u:
+                    cond = {
+                        "cop": "<=",
+                        "args": [cmd.args[0].reg_name, u],
+                    }
+            out = {
+                "block": "if",
+                "condition": cond,
+                "true_branch": [assign_cop([arg_to_bit(cmd.bits[0])], [1])],
+            }
+        case tk.MultiBitOp():
+            # determine number of register operands involved in the operation
+            operand_count = (
+                len(cmd.args) // len(cmd.bits) - 1
+                if op.basic_op.type == tk.OpType.ExplicitPredicate
+                else len(cmd.args) // len(cmd.bits)
+            )
+            out = assign_cop(
+                # Converting to regwise operations that pecos can handle
+                [cmd.bits[0].reg_name],
+                [
+                    {
+                        "cop": cop_from_op_name(op.basic_op.get_name()),
+                        "args": [arg.reg_name for arg in args[:operand_count]],
```

Comment:
I chose to convert this op into a register-wise operation to be handled by PECOS. I believe @qciaran once mentioned PECOS supports operations on unevenly sized registers.

Reply:
If this is the only way to generate from pytket then this works, but still feels a bit fragile. In principle the bits could come from a mixture of "registers".

Reply:
That's fair. Given how obscure these operations seemed, I wasn't sure how much we wanted to investigate this. Do you happen to have a test case for the mixed scenario?

Reply:
OK, here is a slightly contrived way to generate a circuit with two multi-bit AND operations, the first acting registerwise and the second acting on the bits in a different order:
```python
c = Circuit()
areg = c.add_c_register("a", 2)
breg = c.add_c_register("b", 2)
creg = c.add_c_register("c", 2)
c.add_c_and_to_registers(areg, breg, creg)
mbop = c.get_commands()[0].op
c.add_gate(mbop, [areg[0], areg[1], breg[0], breg[1], creg[0], creg[1]])

print(c.get_commands())
```

Output:

```
[AND (*2) a[0], b[0], c[0], a[1], b[1], c[1];, AND (*2) a[0], a[1], b[0], b[1], c[0], c[1];]
```

Reply:
I note, however, that the result of `circuit_to_qasm_str(c, header="hqslib1")` for this circuit is incorrect!
```
OPENQASM 2.0;
include "hqslib1.inc";

creg a[2];
creg b[2];
creg c[2];
c = a & b;
b = a & a;
```

which is a bug in the QASM converter.

https://github.com/CQCL/tket/issues/1327

Reply:
I think it would be OK to simply disallow this case -- but we probably ought to at least check for it.

Reply:
Letting @Asa-Kosto-QTM take a look for disallowing this scenario. Thanks for the cautionary review.

Reply:
Ended up implementing the translation for the example you shared with loads of `zip`s. Hope that suffices for the common use cases.